File: app/src/main/java/za/org/grassroot2/services/SyncOfflineDataService.java
package za.org.grassroot2.services;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;
import timber.log.Timber;
import za.org.grassroot2.GrassrootApplication;
import za.org.grassroot2.R;
import za.org.grassroot2.database.DatabaseService;
import za.org.grassroot2.model.UploadResult;
import za.org.grassroot2.model.alert.LiveWireAlert;
import za.org.grassroot2.model.exception.ServerErrorException;
import za.org.grassroot2.model.network.Syncable;
import za.org.grassroot2.model.request.MemberRequest;
import za.org.grassroot2.model.task.Meeting;
import za.org.grassroot2.model.task.Task;

/**
 * Created by qbasso on 23.10.2017.
 */

public class SyncOfflineDataService extends Service {

    @Inject DatabaseService dbService;
    @Inject NetworkService  networkService;
    private CompositeDisposable disposable = new CompositeDisposable();

    @Override
    public void onCreate() {
        super.onCreate();
        ((GrassrootApplication) getApplication()).getAppComponent().inject(this);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        disposable.add(Observable.zip(dbService.getObjectsToSync(Meeting.class),
                dbService.getObjectsToSync(LiveWireAlert.class),
                dbService.getMemberRequestsToSync(),
                (meetings, alerts, memberRequests) -> combineAndSortSyncables(meetings, alerts, memberRequests)).flatMapIterable(syncables -> syncables).flatMap(syncable -> {
            Timber.d("Syncing object: %s", syncable.toString());
            if (syncable instanceof MemberRequest) {
                handleMemberRequestSync((MemberRequest) syncable);
            } else if (syncable instanceof LiveWireAlert) {
                handleAlertSync((LiveWireAlert) syncable);
            } else {
                handleMeetingSync((Meeting) syncable);
            }
            return Observable.just(true);
        }).subscribeOn(Schedulers.single()).subscribe(o -> stopSelf(), throwable -> {
            Timber.d("Object not synced");
            throwable.printStackTrace();
            stopSelf();
        }));
        return super.onStartCommand(intent, flags, startId);
    }

    private void handleAlertSync(LiveWireAlert syncable) {
        UploadResult result = networkService.uploadEntity(syncable, false).blockingFirst();
        if (result.getUploadException() instanceof ServerErrorException) {
            dbService.delete(LiveWireAlert.class, syncable);
            postFailureMessage(syncable);
        } else if (!TextUtils.isEmpty(result.getServerUid())) {
            syncable.setServerUid(result.getServerUid());
            syncable.setSynced(true);
            dbService.storeObject(LiveWireAlert.class, syncable);
        } else {
            syncable.setSynced(false);
            dbService.storeObject(LiveWireAlert.class, syncable);
        }
    }

    private void postFailureMessage(Syncable syncable) {
        EventBus.getDefault().post(new ObjectOutOfSyncEvent(syncable, buildErrorMessage(syncable)));
    }

    @SafeVarargs
    @NonNull
    private final List<Syncable> combineAndSortSyncables(List<? extends Syncable>... args) {
        List<Syncable> result = new ArrayList<>();
        for (List<? extends Syncable> l : args) {
            result.addAll(l);
        }
        Collections.sort(result, (o1, o2) -> o1.createdDate() > o2.createdDate() ? 1 : (o1.createdDate() < o2.createdDate() ? -1 : 0));
        return result;
    }

    private void handleMemberRequestSync(MemberRequest syncable) {
        Response<Void> response = networkService.inviteContactsToGroup(syncable.groupUid, Collections.singletonList(syncable)).blockingFirst();
        if (!response.isSuccessful()) {
            postFailureMessage(syncable);
        }
        dbService.delete(syncable);
    }

    private void handleMeetingSync(Meeting syncable) throws Exception {
        if (System.currentTimeMillis() >= syncable.getDeadlineMillis()) {
            Timber.d("Deleting object due to start date in the past");
            dbService.delete(Meeting.class, syncable);
            postFailureMessage(syncable);
        }
        Resource<Task> task = networkService.createTask(syncable).blockingFirst();
        if (task.status == Status.SUCCESS) {
            dbService.delete(Meeting.class, syncable);
        } else if (task.status == Status.SERVER_ERROR) {
            dbService.delete(Meeting.class, syncable);
            postFailureMessage(syncable);
        } else if (task.status == Status.ERROR){
            throw new Exception(task.message);
        }
    }

    private String buildErrorMessage(Syncable syncable) {
        if (syncable instanceof MemberRequest) {
            return getString(R.string.invite_failed, ((MemberRequest) syncable).displayName);
        } else if (syncable instanceof LiveWireAlert) {
            return getString(R.string.alert_failed, ((LiveWireAlert) syncable).getHeadline());
        } else {
            if (System.currentTimeMillis() >= ((Meeting)syncable).getDeadlineMillis()) {
                return getString(R.string.meeting_out_of_sync, ((Meeting) syncable).getName());
            } else {
                return getString(R.string.meeting_failed, ((Meeting) syncable).getName());
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposable.dispose();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    public static class ObjectOutOfSyncEvent {
        public final Syncable syncable;
        @Nullable
        public final String   msg;

        public ObjectOutOfSyncEvent(Syncable syncable, @Nullable String msg) {
            this.syncable = syncable;
            this.msg = msg;
        }
    }
}

File: app/src/main/java/za/org/grassroot2/services/SyncOfflineDataService.kt
package za.org.grassroot2.services

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.text.TextUtils
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Function3
import io.reactivex.schedulers.Schedulers
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import za.org.grassroot2.GrassrootApplication
import za.org.grassroot2.R
import za.org.grassroot2.database.DatabaseService
import za.org.grassroot2.model.alert.LiveWireAlert
import za.org.grassroot2.model.exception.ServerErrorException
import za.org.grassroot2.model.network.Syncable
import za.org.grassroot2.model.request.MemberRequest
import za.org.grassroot2.model.task.Meeting
import java.util.ArrayList
import javax.inject.Inject
import kotlin.Comparator
import kotlin.Exception
import kotlin.Int
import kotlin.String

/**
 * Created by qbasso on 23.10.2017.
 */

class SyncOfflineDataService : Service() {

    @Inject internal lateinit var dbService: DatabaseService
    @Inject internal lateinit var networkService: NetworkService

    private val disposable = CompositeDisposable()

    override fun onCreate() {
        super.onCreate()
        (application as GrassrootApplication).appComponent.inject(this)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val meetingObservable: Observable<List<Meeting>> = dbService.getObjectsToSync(Meeting::class.java)
        val alertObservable: Observable<List<LiveWireAlert>> = dbService.getObjectsToSync(LiveWireAlert::class.java)
        val memberRequestObservable: Observable<List<MemberRequest>> = dbService.getMemberRequestsToSync()

        val combinedObservable = Observable.zip<List<Meeting>, List<LiveWireAlert>, List<MemberRequest>, List<Syncable>>(
                meetingObservable, alertObservable, memberRequestObservable, Function3 { meetings: List<Meeting>, alerts: List<LiveWireAlert>, requests: List<MemberRequest> ->
                    combineAndSortSyncables(meetings, alerts, requests)});

        disposable.add(combinedObservable
                .flatMapIterable { syncables -> syncables }
                .flatMap { syncable ->
                    Timber.d("Syncing object: %s", syncable.toString())
                    if (syncable is MemberRequest) {
                        handleMemberRequestSync(syncable)
                    } else if (syncable is LiveWireAlert) {
                        handleAlertSync(syncable)
                    } else {
                        handleMeetingSync(syncable as Meeting)
                    }
                    Observable.just(true) }
                .subscribeOn(Schedulers.single())
                .subscribe({ o -> stopSelf() }, { throwable ->
                    Timber.d("Object not synced")
                    throwable.printStackTrace()
                    stopSelf()
                }))
        return super.onStartCommand(intent, flags, startId)
    }

    private fun handleAlertSync(syncable: LiveWireAlert) {
        val result = networkService.uploadEntity(syncable, false).blockingFirst()
        if (result.uploadException is ServerErrorException) {
            dbService.delete(LiveWireAlert::class.java, syncable)
            postFailureMessage(syncable)
        } else if (!TextUtils.isEmpty(result.serverUid)) {
            syncable.serverUid = result.serverUid
            syncable.isSynced = true
            dbService.storeObject(LiveWireAlert::class.java, syncable)
        } else {
            syncable.isSynced = false
            dbService.storeObject(LiveWireAlert::class.java, syncable)
        }
    }

    private fun postFailureMessage(syncable: Syncable) {
        EventBus.getDefault().post(ObjectOutOfSyncEvent(syncable, buildErrorMessage(syncable)))
    }

    @SafeVarargs
    private fun combineAndSortSyncables(vararg args: List<Syncable>): List<Syncable> {
        val result = ArrayList<Syncable>()
        for (l in args) {
            result.addAll(l)
        }
        result.sortWith(Comparator { o1, o2 -> if (o1.createdDate() > o2.createdDate()) 1 else if (o1.createdDate() < o2.createdDate()) -1 else 0 })
        return result
    }

    private fun handleMemberRequestSync(syncable: MemberRequest) {
        val response = networkService.inviteContactsToGroup(syncable.groupUid, listOf(syncable)).blockingFirst()
        if (!response.isSuccessful) {
            postFailureMessage(syncable)
        }
        dbService.delete(syncable)
    }

    @Throws(Exception::class)
    private fun handleMeetingSync(syncable: Meeting) {
        if (System.currentTimeMillis() >= syncable.deadlineMillis) {
            Timber.d("Deleting object due to start date in the past")
            dbService.delete(Meeting::class.java, syncable)
            postFailureMessage(syncable)
        }
        val task = networkService.createTask(syncable).blockingFirst()
        if (task.status == Status.SUCCESS) {
            dbService.delete(Meeting::class.java, syncable)
        } else if (task.status == Status.SERVER_ERROR) {
            dbService.delete(Meeting::class.java, syncable)
            postFailureMessage(syncable)
        } else if (task.status == Status.ERROR) {
            throw Exception(task.message)
        }
    }

    private fun buildErrorMessage(syncable: Syncable): String {
        return if (syncable is MemberRequest) {
            getString(R.string.invite_failed, syncable.displayName)
        } else if (syncable is LiveWireAlert) {
            getString(R.string.alert_failed, syncable.headline)
        } else {
            if (System.currentTimeMillis() >= (syncable as Meeting).deadlineMillis) {
                getString(R.string.meeting_out_of_sync, syncable.name)
            } else {
                getString(R.string.meeting_failed, syncable.name)
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    class ObjectOutOfSyncEvent(val syncable: Syncable, val msg: String?)
}
